src/lib.rs

```diff
@@ -1851,7 +2438,205 @@ impl GooseAttack {
             }
             message = metric_receiver.try_recv();
         }
-        received_message
+        Ok(received_message)
+    }
+}
+
+/// All run-time options can optionally be configured with custom defaults. For
+/// example, you can optionally configure a default host for the load test. This is
+/// used if no per-GooseTaskSet host is defined, no `--host` CLI option is
+/// configured, and if the GooseTask itself doesn't hard-code the host in the base
+/// url of its request. In that case, this host is added to all requests.
+///
+/// For example, a load test could be configured to default to running against a local
+/// development container, and the `--host` option could be used to override the host
+/// value to run the load test against the production environment.
+///
+/// # Example
+/// ```rust,no_run
+///     use goose::prelude::*;
+///
+/// fn main() -> Result<(), GooseError> {
+///     GooseAttack::initialize()?
+///         .set_default(GooseDefault::Host, "local.dev");
+///
+///     Ok(())
+/// }
+/// ```
+///
+/// The following run-time options can be configured with a custom default using a
+/// borrowed string slice (`&str`):
+///  - GooseDefault::Host
+///  - GooseDefault::LogFile
+///  - GooseDefault::MetricsFile
+///  - GooseDefault::MetricsFormat
+///  - GooseDefault::DebugFile
+///  - GooseDefault::DebugFormat
+///  - GooseDefault::ManagerBindHost
+///  - GooseDefault::ManagerHost
+///
+/// The following run-time options can be configured with a custom default using a
+/// `usize` integer:
+///  - GooseDefault::Users
+///  - GooseDefault::HatchRate
+///  - GooseDefault::RunTime
+///  - GooseDefault::LogLevel
+///  - GooseDefault::Verbose
+///  - GooseDefault::ThrottleRequests
+///  - GooseDefault::ExpectWorkers
+///  - GooseDefault::ManagerBindPort
+///  - GooseDefault::ManagerPort
+///
+/// The following run-time flags can be configured with a custom default using a
+/// `bool` (and otherwise default to `false`).
+///  - GooseDefault::OnlySummary
+///  - GooseDefault::NoResetMetrics
+///  - GooseDefault::NoMetrics
+///  - GooseDefault::NoTaskMetrics
+///  - GooseDefault::StatusCodes
+///  - GooseDefault::StickyFollow
+///  - GooseDefault::Manager
+///  - GooseDefault::NoHashCheck
+///  - GooseDefault::Worker => panic!(format!(
+///
+/// # Another Example
+/// ```rust,no_run
+///     use goose::prelude::*;
+///
+/// fn main() -> Result<(), GooseError> {
+///     GooseAttack::initialize()?
+///         .set_default(GooseDefault::OnlySummary, true)
+///         .set_default(GooseDefault::Verbose, 1)
+///         .set_default(GooseDefault::MetricsFile, "goose-metrics.log");
+///
+///     Ok(())
+/// }
+/// ```
+pub trait GooseDefaultType<T> {
```

Comment:
@LionsAd Here's where we use a Generic `GooseDefaultType`, and then implement for `&str`, `usize` and `bool`. In order to have friendly error messages, we have to handle the entire `GooseDefault` enum in all three implementations.